### PR TITLE
wip chain registry 

### PR DESCRIPTION
**Pull request by Deneath**

### CODE REVIEW

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/di/WalletFeatureModule.kt [hunk 1]
```diff
@@ -241,11 +229,27 @@ class WalletFeatureModule {
         preferences,
         selectedFiat,
         updatesMixin,
-        xcmService,
-        currentAccountAddressUseCase,
         xcmEntitiesFetcher
     )
 
+    @Provides
+    @Singleton
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/di/WalletFeatureModule.kt:236]: `Singleton` redundant for `Interactor`

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+class XcmInteractor(
+    private val walletInteractor: WalletInteractor,
+    private val chainRegistry: ChainRegistry,
+    private val currentAccountAddress: CurrentAccountAddressUseCase,
+    private val xcmEntitiesFetcher: XcmEntitiesFetcher,
+    private val accountInteractor: AccountInteractor,
+    private val runtimeFilesCache: RuntimeFilesCache
+) {
+
+    private val xcmService: MutableStateFlow<XcmService?> = MutableStateFlow(null)
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt:40]: We can move `xcmService` to `XcmRepository` to keep Interactor stateless

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    private val nonNullXcmService = xcmService.filterNotNull()
+
+    suspend fun initXcmService(originChainId: ChainId, destinationChainId: ChainId) {
+        xcmService.value = null
+        val fromChainData = buildFromChainData(originChainId)
+        val destinationChainMetadata = runtimeFilesCache.getChainMetadata(destinationChainId)
+        this.xcmService.value = XcmService.create(fromChainData, destinationChainId, destinationChainMetadata)
+    }
+
+    private suspend fun buildFromChainData(originChainId: ChainId): FromChainData {
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt:50]: We can move this logic to `XcmRepository`, but it seems too hard

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    suspend fun performCrossChainTransfer(
+        transfer: CrossChainTransfer,
+        fee: BigDecimal,
+        tipInPlanks: BigInteger?
+    ): Result<String> {
+        return runCatching {
+            val originChain = chainRegistry.getChain(transfer.originChainId)
+            val destinationChain = chainRegistry.getChain(transfer.destinationChainId)
+            val selfAddress = currentAccountAddress(originChain.id) ?: throw IllegalStateException("No self address")
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt:94]: `throw IllegalStateException` -> `error`

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                amount = transfer.fullAmountInPlanks
+            )
+        }
+    }
+
+    suspend fun getXcmDestFee(
+        destinationChainId: ChainId,
+        tokenSymbol: String
+    ): BigDecimal? {
+        val r = runCatching {
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt:111]: `r`?

File: feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        return r.getOrNull()
+    }
+
+    suspend fun getXcmOrigFee(
+        originNetworkId: ChainId,
+        destinationNetworkId: ChainId,
+        asset: CoreAsset,
+        amount: BigDecimal
+    ): BigDecimal? {
+        val r = runCatching {
```
Comment (VitalyPeryatin) [inline feature-wallet-impl/src/main/java/jp/co/soramitsu/wallet/impl/domain/XcmInteractor.kt:127]: `r`?

**Review by VitalyPeryatin — CHANGES_REQUESTED** [review]

**Review by PankraSerg — APPROVED** [review]

**Review by PankraSerg — APPROVED** [review]

**Review by VitalyPeryatin — APPROVED** [review]

**Review by VitalyPeryatin — APPROVED** [review]

**Pull request merged**